packages/lib/src/input.ts

```diff
@@ -0,0 +1,209 @@
+import { flags } from '@oclif/command'
+import fs from 'fs'
+import path from 'path'
+import yaml from 'js-yaml'
+import util from 'util'
+
+import { logManager } from './logger'
+import { SmartThingsCommand } from './smartthings-command'
+
+
+export const readFile = util.promisify(fs.readFile)
+
+export enum IOFormat {
+	YAML = 'yaml',
+	JSON = 'json',
+
+	// for input, this is Q & A or command line, for output, it's a human-readable table format
+	COMMON = 'common',
+}
+
+export function formatFromFilename(filename: string): IOFormat {
+	const ext = path.extname(filename).toLowerCase()
+	if (ext === '.yaml' || ext === '.yml') {
+		return IOFormat.YAML
+	}
+	if (ext === '.json') {
+		return IOFormat.JSON
+	}
+	logManager.getLogger('cli').warn(`could not determine file type from filename "${filename}, assuming YAML`)
+	return IOFormat.YAML
+}
+
+// Flags common to both input and output.
+export const commonIOFlags = {
+	indent: flags.integer({
+		description: 'specify indentation for formatting JSON or YAML output',
+	}),
+	json: flags.boolean({
+		description: 'use JSON format of input and/or output',
+		char: 'j',
+	}),
+	yaml: flags.boolean({
+		char: 'y',
+		description: 'use YAML format of input and/or output',
+	}),
+}
+
+export const inputFlag = {
+	input: flags.string({
+		char: 'i',
+		description: 'specify input file',
+	}),
+}
+
+
+export function parseJSONOrYAML<T>(rawInputData: string, source: string): T {
+	const data = yaml.safeLoad(rawInputData)
+	if (!data) {
+		throw Error(`did not get any data from ${source}`)
+	}
+	if (typeof data === 'string') {
+		throw Error(`got simple string from ${source}`)
+	}
+	return data as unknown as T
+}
+
+export function readDataFromStdin(): Promise<string> {
+	return new Promise((resolve, reject) => {
+		try {
+			const stdin = process.stdin
+			const inputChunks: string[] = []
+			stdin.resume()
+			stdin.on('data', chunk => {
+				inputChunks.push(chunk.toString())
+			})
+			stdin.on('end', () => {
+				resolve(inputChunks.join(''))
+			})
+		} catch (error) {
+			reject(error)
+		}
+	})
+}
+
+
+export interface InputProcessor<T> {
+	/**
+	 * Return the type of input this processor retrieved. This should not be called until after
+	 * data has been read via `read` so throwing an exception before that is acceptable.
+	 */
+	readonly ioFormat: IOFormat
+
+	/**
+	 * This is called before calling `read` to determine if there is input from this processor.
+	 * The processor should return `false` to indicate `read` should not be called or `true`
+	 * if calling `read` can be expected to return data.
+	 */
+	hasInput(): boolean
```

Comment:
Nothing here is really keeping me from just calling `read` anyway. Is this providing any benefit besides gate keeping the call to `read`? Another pattern could be to simply try reading from the source and handling any errors that would arise from the processor not being "ready". I think I see us [checking again anyway](https://github.com/SmartThingsCommunity/smartthings-cli/pull/117/files#diff-3c4a91599b07bd690b22afdcc40f5e76b994f3d264e59baf12877f582f36fc51R120) in some `read` calls too.

Reply:
Yeah, the point of this is to report whether the particular processor can provide input. I think it's important to keep the logic of "having input" (`hasInput`) and "providing input" (`read`) together since each input source has its own unique way of determining that. Tying them together in an interface seems perfectly natural to me.